docs/design/master0-recovery.md

```diff
@@ -0,0 +1,111 @@
+# Master0 recovery design
+
+## Motivations
+
+Master0 plays an important role in sealer cluster, it hosts many important components such as sealer builtin registry,
+application images launched history and others cluster configuration files. if the master0 node is abnormal, a highly
+available solution is required to recover it on other master nodes.
+
+Currently, we have plans to save clusterfile to the cluster. while for built registry ,it is still deployed a single
+instance on master0 node. So the following document are mainly for built registry.
+
+## Proposal
+
+### For builtin registry
+
+* expose registry configuration to user thought clusterfile.
+* distribute all registry data split from cluster image registry directory to each deploy host, whether it is a cluster
+  image or an application image.
+* save all registry configuration generated form sealer such as, basic auth information and https certificate.
+* save the boot history of each application image information such as tag,version and image id to cluster, which was
+  launched through the sealer run-app.
+
+## Implementation
+
+### For configuration of builtin registry
+
+1. expose below `Registry` struct to clusterfile which is contains configurations about local registry and remote
+   registry.
+
+```yaml
+type Registry struct {
+  Domain        string        `json:"domain,omitempty"`
+  Port          int           `json:"port,omitempty"`
+  Username      string        `json:"username,omitempty"`
+  Password      string        `json:"password,omitempty"`
+  LocalRegistry LocalRegistry `json:"local_registry,omitempty"`
+}
+  type LocalRegistry struct {
+  // DeployHosts is the target host list that local registry will be deployed on.
+  // if not set ,master0 will be the default value.
+  DeployHosts []net.IP `json:"deploy_hosts,omitempty"`
+  // InsecureMode indicated that whether the local registry is exposed in HTTPS.
+  // if true sealer will generate default ssl cert.
+  InsecureMode bool `json:"insecure_mode,omitempty,omitempty"`
+}
+```
+
+2. save registry configuration to cluster as configmap with name "sealer-registry" in namespace "kub-system".
```

Comment:
use secret not configmap.

kube-system

Reply:
we can also save to others master`s local disk at the same time

Reply:
no need do this. k8s use etcd to provide HA storage.